corehq/apps/reports/analytics/esaccessors.py

```diff
@@ -341,34 +341,32 @@ def get_form_counts_by_user_xmlns(domain, startdate, enddate, user_ids=None,
     es_instance = ES_EXPORT_INSTANCE if export else ES_DEFAULT_INSTANCE
     query = (FormES(es_instance_alias=es_instance)
              .domain(domain)
-             .filter(date_filter_fn(gte=startdate, lt=enddate))
-             .aggregation(
-                 TermsAggregation('user_id', 'form.meta.userID').aggregation(
-                     TermsAggregation('app_id', 'app_id').aggregation(
-                         TermsAggregation('xmlns', 'xmlns')
-                     )
-                 )
-             )
-             .size(0)
-    )
+             .filter(date_filter_fn(gte=startdate, lt=enddate)))
 
     if user_ids:
         query = (query
             .user_ids_handle_unknown(user_ids)
             .remove_default_filter('has_user'))
         missing_users = None in user_ids
-        if missing_users:
-            query = query.aggregation(
-                MissingAggregation('missing_user_id', 'form.meta.userID').aggregation(
-                    TermsAggregation('app_id', 'app_id').aggregation(
-                        TermsAggregation('xmlns', 'xmlns')
-                    )
-                )
-            )
 
     if xmlnss:
         query = query.xmlns(xmlnss)
 
+    query = query.aggregation(
+        TermsAggregation('user_id', 'form.meta.userID').aggregation(
+            TermsAggregation('app_id', 'app_id').aggregation(
+                TermsAggregation('xmlns', 'xmlns')
+            )
+        )).size(0)
+    if user_ids and missing_users:
+        query = query.aggregation(
+            MissingAggregation('missing_user_id', 'form.meta.userID').aggregation(
+                TermsAggregation('app_id', 'app_id').aggregation(
+                    TermsAggregation('xmlns', 'xmlns')
+                )
+            )
+        )
```

Comment:
changing the order that these calls happen in won't make a difference. This just updates the query dict which then get's compiled and executed below `query.run()`. ES will do the filtering prior to the aggregation stage.

Reply:
Yeah, I realized, and was trying to add FilterAggregation which I thought limits the scope, but wasn't helping. I am considering an approach where if the count of docs is very low after filtering, to do the agg in Python, I will do that in a separate PR. For now, I will revert this particular agg change

Reply:
Reverted that change, now this PR is just about timeout

Reply:
what's the problem you're trying to solve? Doing an agg in ES on a small number of docs shouldn't be an issue.

Reply:
The aggregations over these fields even when the result set is very small makes that field data for all docs into fielddata memory/cache. So, aggregating in Python when total count of docs after limiting to ~1000 docs would prevent that.

If you are aware of any better way to limit the scope of agg, that will be great. I will do some more reading anyway to see if there is any.

Reply:
yea, ok, that's the same issue as sorting.

Reply:
Hmm, actually the query we have generates a filtered query, per [these docs](https://www.oreilly.com/library/view/elasticsearch-the-definitive/9781449358532/part04ch06.html) and this [SO answer](https://stackoverflow.com/questions/28258481/spring-data-elastisearch-aggregations-with-filters/28263684#28263684) the aggs should get applied on filtered docs only. Though, I don't understand why it takes so much more time to do aggregations on even ~10 docs whereas getting the raw hits was very fast.